internal/keepers/util.go

```diff
@@ -129,16 +131,21 @@ func shuffledDedupedKeyList(attributed []types.AttributedObservation, key [16]by
 		}
 
 		var ob []ktypes.UpkeepKey
-		err = decode(b, &ob)
 
 		// a single observation returning an error here can void all other
 		// good observations. ensure this loop continues on error, but collect
 		// them and throw an error if ALL observations fail at this point.
+		var keys []chain.UpkeepKey
```

Comment:
No change, but this is going to come back to bite us. Referencing the distinct struct type here instead of an interface makes the code less adaptable to different chain implementations in the future. I think the only choice we have though is to build the structure into the observation encoding/decoding to decide on decoded struct types. This will need to be a TODO for later.